File: Assets/Scripts/PlayerController.cs
using System.Collections;
using System.Collections.Generic;
using DG.Tweening;
using UnityEngine;

public class PlayerController : MonoBehaviour
{
    public Rigidbody2D rigidBody2D;
    public Animator animator;
    public GameManager gameManager;
    public GameObject ballPrefab;
    public Transform firePosition;
    public float rotateCoefficient = 1f;
    public float flyForce = 6f;
    public Transform playerRendererTransform;
    public AudioController audioController;
    
    private enum PlayerState
    {
        Idle,   // 0
        UpDown, // 1
        Fly     // 2
    }

    // Update is called once per frame
    void Update()
    {   
        if(!gameManager.isGameStart)return;
        if (Input.GetMouseButtonDown(0) || Input.GetKeyDown(KeyCode.C))
        {
             Fly();
             CreateBall();
        }
        playerRendererTransform.DORotateQuaternion(Quaternion.Euler(0, 0, rigidBody2D.velocity.y * rotateCoefficient), 0.3f);
    }

    public void Fly()
    {   
        audioController.PlaySfx(audioController.jump);
        rigidBody2D.velocity = new Vector2(0, flyForce);
    }
    
    public void CreateBall()
    {
        // 生成球的随机位置偏移
        Vector2 positionOffset = new Vector2(
            Random.Range(-0.2f, 0.2f), // X 轴偏移（-0.1 到 0.1）
            Random.Range(-0.2f, 0.2f)  // Y 轴偏移（-0.1 到 0.1）
        );

        // 生成球的随机旋转角度
        float rotationOffset = Random.Range(-30f, 30f); // 旋转角度偏移

        // 计算最终生成位置和旋转
        Vector2 spawnPosition = firePosition.position + (Vector3)positionOffset;
        Quaternion spawnRotation = Quaternion.Euler(0, 0, rotationOffset);

        // 实例化球
        GameObject ball = Instantiate(ballPrefab, spawnPosition, spawnRotation);
        ball.SetActive(true);
    }

    public void ChangeState(bool isFly, bool isSimulated = false)
    {
        if (isFly)
        {   
            animator.SetInteger("state", (int)PlayerState.Fly);
        }
        else
        {
            animator.SetInteger("state", (int)PlayerState.UpDown);

        }
        rigidBody2D.simulated = isSimulated;
    }

}

File: Assets/Lua/Game/Player/PlayerController.lua
PlayerController = {}

-- 属性
local rigidBody2D = nil
local animator = nil
local gameManager = nil
local ballPrefab = nil
local firePosition = nil
local playerRendererTransform = nil
local audioController = nil
local rotateCoefficient = 1.0
local flyForce = 6.0

-- 玩家状态枚举
local PlayerState = {
    Idle = 0,
    UpDown = 1,
    Fly = 2
}

-- 初始化函数
function PlayerController.Start(go)
    local playerGO = go or CS.UnityEngine.GameObject.Find("Player")
    
    -- 获取组件引用
    rigidBody2D = playerGO.GetComponent(typeof(CS.UnityEngine.Rigidbody2D))
    animator = playerGO.GetComponent(typeof(CS.UnityEngine.Animator))
    playerRendererTransform = playerGO.transform.Find("Renderer")
    
    -- 获取其他对象引用
    gameManager = CS.UnityEngine.GameObject.Find("GameManager"):GetComponent(typeof(CS.GameManager))
    audioController = CS.UnityEngine.GameObject.Find("AudioManager"):GetComponent(typeof(CS.AudioController))
    
    -- 获取预制体和生成位置
    local playerController = playerGO:GetComponent(typeof(CS.PlayerController))
    if playerController then
        ballPrefab = playerController.ballPrefab
        firePosition = playerController.firePosition
        rotateCoefficient = playerController.rotateCoefficient
        flyForce = playerController.flyForce
    end
end

-- 更新函数
function PlayerController.Update()
    if not gameManager.isGameStart then
        return
    end
    
    if CS.UnityEngine.Input.GetMouseButtonDown(0) or CS.UnityEngine.Input.GetKeyDown(CS.UnityEngine.KeyCode.C) then
        PlayerController.Fly()
        PlayerController.CreateBall()
    end
    
    if playerRendererTransform then
        -- 使用 DOTween 进行旋转，但我们在 Lua 中使用简单的方式模拟
        local targetRotation = CS.UnityEngine.Quaternion.Euler(0, 0, rigidBody2D.velocity.y * rotateCoefficient)
        playerRendererTransform.rotation = CS.UnityEngine.Quaternion.Lerp(
            playerRendererTransform.rotation, 
            targetRotation, 
            0.3
        )
    end
end

-- 飞行函数
function PlayerController.Fly()
    audioController:PlaySfx(audioController.jump)
    rigidBody2D.velocity = CS.UnityEngine.Vector2(0, flyForce)
end

-- 创建小球
function PlayerController.CreateBall()
    -- 生成球的随机位置偏移
    local positionOffset = CS.UnityEngine.Vector2(
        CS.UnityEngine.Random.Range(-0.2, 0.2),
        CS.UnityEngine.Random.Range(-0.2, 0.2)
    )
    
    -- 生成球的随机旋转角度
    local rotationOffset = CS.UnityEngine.Random.Range(-30, 30)
    
    -- 计算最终生成位置和旋转
    local spawnPosition = firePosition.position + CS.UnityEngine.Vector3(positionOffset.x, positionOffset.y, 0)
    local spawnRotation = CS.UnityEngine.Quaternion.Euler(0, 0, rotationOffset)
    
    -- 实例化球
    local ball = CS.UnityEngine.Object.Instantiate(ballPrefab, spawnPosition, spawnRotation)
    ball:SetActive(true)
end

-- 改变状态
function PlayerController.ChangeState(isFly, isSimulated)
    isSimulated = isSimulated or false
    
    if isFly then
        animator:SetInteger("state", PlayerState.Fly)
    else
        animator:SetInteger("state", PlayerState.UpDown)
    end
    
    rigidBody2D.simulated = isSimulated
end

return PlayerController 